### ISSUE TITLE
Bugzilla Crawler

### ISSUE THREAD

**Issue by leilani-reich** [sailuh/kaiaulu#153]
### Proposal

We will implement a Bugzilla crawler that will be directly called from R. Bugzilla has a REST API, so we can parse the JSON that is returned by the API call in the R function.

The main contributions for this issue will be the Bugzilla crawler functions, unit tests for these functions, and a vignette demonstrating how to use the Bugzilla crawler.

### Architecture

NOTE: We should discuss the architecture of the Bugzilla crawler before starting any code.

As stated by Carlos:
The issues raised on issue #70 of PERCEIVE provide good motivation on why we should refine the requirements of it upfront: https://github.com/sailuh/perceive/issues/70. 

More specifically, we will want to break the Bugzila crawler into a 
- i) "downloader as-is" (which I believe the raw data is json)
- ii) parser of said json into tables. 

The downloader would be under R/download.R, where as the crawler under R/parser.R.

**Comment by carlosparadis** [comment]
This may prove useful:

Possibly useful Docs:
- https://wiki.mozilla.org/Bugzilla:REST_API
- https://bz.apache.org/ooo/docs/en/html/integrating/api/Bugzilla/WebService/Server/REST.html

E.g. Bug 35: https://bz.apache.org/bugzilla/show_bug.cgi?id=35
Bug 35 via REST interface: https://bz.apache.org/bugzilla/rest.cgi/bug/35

Also, please let me know which R libraries you believe are needed for this task. Ideally, the ones already listed as dependencies should suffice:

https://github.com/sailuh/kaiaulu/blob/c968a6ded22610db52e9aaa261d258d5d7128976/DESCRIPTION#L34-L45 

R/download.R is a good start for examples using them for downloading data, and R/parse.R for parsing input from Perceval via system2 calls.

**Comment by leilani-reich** [comment]
Hi, here's a rough function for the downloader. It get's the data using the REST API, and will start bug retrieval based on the date parameter. There is an option to download comments for the bugs if the user wants. The function will output a list from which users can obtain issue data and comment data in json format.

```r
#'  Download Bugzilla issues and comments using the Bugzilla REST API
#'
#' @param bugzilla_site link to specific bugzilla site
#' @param date the date to start bug retrieval
#' @param comments whether to download comments
#' @export
download_bugzilla_from_rest_api <- function(bugzilla_site, date, comments=FALSE){
  # Format link to retrieve data using Bugzilla REST API
  bugzilla_site <- paste(bugzilla_site, "/rest", sep="")
  
  # Get the issues starting from the specified date
  issues <- GET(paste(bugzilla_site, "/bug", "?creation_time=", date, sep=""))
  json_issues <- rawToChar(issues$content)
  
  # Get the bug ids
  bug_issue_table <- data.table(jsonlite::stream_in(textConnection(json_issues),verbose = FALSE))
  bug_ids <- bug_issue_table[["bugs"]][[1]][["id"]]
  print(bug_ids)
  
  # Download the comments if user wants them
  json_comments <- vector()
  if (comments == TRUE){
    # Get the comments associated with the bugs acquired
    for (bug in bug_ids){
      comments <- GET(paste(bugzilla_site, "/bug/", bug, "/comment", "?creation_time=", date, sep=""))
      json_comments <- c(json_comments, rawToChar(comments$content))
    }
  }
  
  json_issues_comments <- list()
  json_issues_comments[["issues"]] <- jsonlite::toJSON(jsonlite::fromJSON(json_issues))
  json_issues_comments[["comments"]] <- jsonlite::toJSON(lapply(json_comments, jsonlite::fromJSON))
  
  return(json_issues_comments)
}
```

Please let me know how I can improve.

Thanks,
Leilani

**Comment by leilani-reich** [comment]
Revisions requested by Carlos on my download bugzilla function:

1. Currently, your function mimics Perceval limitation of only being capable to download from a start_date, but can not specify an end date. Thus, making harder to recover only portions of data at a time. You want to check the REST API for other parameters in bugzilla. 
   - When you do, please also keep an eye out for any endpoints that allow you to specify the issue number range you wish to download (we may need it, I am not certain yet). Please also link on our Kaiaulu issue the manual you are using to read about what endpoints are available if it is not listed there yet.
2. You should use a timestamp instead of date (i.e. day/month/year hh:mm:ss tz) for precise time range. 
3. Currently, your function has 1 get request for downloading all the issues, but given the bugzilla endpoint, let's assume X issues, followed by X get requests to download the comments. This being the case, because you can only request comments of one issue at a time. Your function then consolidates all comments into a file. This is the part where I need to give more thought, but knowing the answer to the above would be helpful.
4. I mentioned you should revise the file naming scheme by checking how it is done for other download functions (we may be inconsistent). Something along the lines of bugzilla_projectname_starttimestamp_endtimestamp.json would be a good candidate.

**Comment by leilani-reich** [comment]
Hi Carlos,
### 1. 
in regards to setting an end timestamp, I explored the [Bugzilla REST API documentation ](https://bugzilla.readthedocs.io/en/latest/api/index.html) and the web for answers, but I could not find any "end_date" parameter to specify, or any way to search with the creation_date ending at a particular time. In the documentation, if you look at the [search bugs](https://bugzilla.readthedocs.io/en/latest/api/core/v1/bug.html#search-bugs) section, for the creation_time parameter, it says "Searches for bugs that were created at this time or later. May not be an array." So I can't specify an ending range for creation_time. At least, not from what I've seen and tried.

However, I have been able to reduce the number of GET calls for my comments by only extracting bug ids, and thus only comments, from the range (including start_timestamp and end_timestamp) that the user specifies to the functions. This is shown in my revised function below.

### 2. 
I've edited my function to have start_timestamp and end_timestamp parameters.

### 3. 
I'm not sure what you were asking for this third point. In regards to GET calls, there is a single GET call for retrieving the issues, and multiple (equivalent to number of bugs) for retrieving the comments.

### 4. 
See my revised function for how I revised the file naming scheme. I saved the files to the ../../rawdata/issues_trackers folder, which was the convention used in the conf files. I also added a project_name parameter so the user could specify the project name for the saved files. I thought of just taking the bugzilla_site the user inputs and extracting a project name from that, but I thought that wouldn't work well, given that bugzilla sites may have names like "http://cvs.alterplast.ru/bugs/".

### 5. 
- Here's a link to what the json from the GET call for the issues looks like https://bugzilla.redhat.com/rest/bug?creation_time=2023-03-01.
- Also, here's a link to what the json for comments from a single bug looks like https://bugzilla.redhat.com/rest/bug/2157211/comment.


```r
#'  Download Bugzilla issues and comments using the Bugzilla REST API
#'
#' @param bugzilla_site link to specific bugzilla site
#' @param start_timestamp when to start bug retrieval (ex. 2023-01-01T00:14:57Z)
#' @param end_timestamp when to end bug retrieval (ex. 2023-01-01T00:14:57Z)
#' @param project_name the name of the bugzilla project (ex. redhat for https://bugzilla.redhat.com")
#' @param comments whether to download comments
#' @export
download_bugzilla_from_rest_api <- function(bugzilla_site, start_timestamp, end_timestamp, project_name, comments=FALSE){
  # Format link to retrieve data using Bugzilla REST API
  bugzilla_site <- paste(bugzilla_site, "/rest", sep="")
  
  # Get the issues starting from the specified date
  issues <- httr::GET(paste(bugzilla_site, "/bug", "?creation_time=", start_timestamp, sep=""))
  json_issues <- rawToChar(issues$content)
  
  # Get table for the bug issues
  bug_issue_table <- data.table(jsonlite::stream_in(textConnection(json_issues),verbose = FALSE))
  
  # Get table in actual table format
  table <- data.table(bug_issue_table[["bugs"]][[1]])
  
  # Filter the table to only include bugs within date range
  filtered_issue_table <- table[table$creation_time <= end_timestamp]
  
  # Get the bug ids. Needed to know which comments to retrieve.
  bug_ids <- filtered_issue_table[["id"]]
  
  # Download the comments if user wants them
  json_comments <- vector()
  if (comments == TRUE){
    # Get the comments associated with the bugs acquired
    for (bug in bug_ids){
      comments <- httr::GET(paste(bugzilla_site, "/bug/", bug, "/comment", sep=""))
      json_comments <- c(json_comments, rawToChar(comments$content))
    }
  }
  
  # List to hold the issues and comments
  json_issues_comments <- list()
  
  json_issues_comments[["issues"]] <- jsonlite::toJSON(filtered_issue_table)
  json_issues_comments[["comments"]] <- jsonlite::toJSON(lapply(json_comments, jsonlite::fromJSON))
  
  
  # Save issues and comments
  jsonlite::write_json(json_issues_comments[["issues"]], paste("../../rawdata/issue_tracker/",  project_name,
                                                                "_", "issues_",
                                                                start_timestamp, "_", end_timestamp, ".json", sep=""))

  jsonlite::write_json(json_issues_comments[["comments"]], paste("../../rawdata/issue_tracker/",  project_name,
                                                                "_", "comments_",
                                                                start_timestamp, "_", end_timestamp, ".json", sep=""))
    
  return(json_issues_comments)
}

# Example of calling the function
project_bugzilla_rest <- download_bugzilla_from_rest_api("https://bugzilla.redhat.com", "2023-01-01", "2023-01-10", project_name="redhat", comments=TRUE)

project_bugzilla_rest_issues <- data.table(jsonlite::stream_in(textConnection(bugzilla_rest_api_json[["issues"]])))

project_bugzilla_rest_comments <- data.table(jsonlite::stream_in(textConnection(bugzilla_rest_api_json[["comments"]]))[["bugs"]])
```

Let me know what you think.

Thanks,
Leilani

**Comment by carlosparadis** [comment]
@leilani-reich @malialiu @nicoelee123 

Since there is a task dependency in Milestone 2, and I believe Nico is still on Milestone 1, can we reverse the assignment between Malia and Nico for Milestone 2? If needed, you can re-balance Milestone 3 tasks accordingly, effort-wise. For example, Nico can have 1+ more functions to make wrappers on Milestone 3 since Milestone 1 and 2 were solely for unit testing, or some of the other tasks involved in the sub-milestones. This would also be important, so Nico has a few commits contributing features to showcase.

I want to ensure that Milestone 2 is completed as a deliverable since a downloader with no parser would be of little use for analysis. At this point, it is preferable that Malia takes over the parser of Bugzilla crawler, and Nico skips the test of Milestone 2 straight into the features of Milestone 3. Then, after some progress is made, Nico can try to catch up afterward on the unit tests of Milestone 2 (however, Nico's Milestone 1 should be finished before Milestone 3).

Does that work for the group? I am open to other arrangements if we can get Milestone 2 feature worked out.

**Comment by leilani-reich** [comment]
@carlosparadis 
Hi, that works for us.

**Comment by carlosparadis** [comment]
Thanks, works for me then! Let's wait on @rnkazman to confirm since he is the one who does the final ok for your capstone :) .

**Comment by rnkazman** [comment]
Yes, this makes sense to me as well.  Thanks!

**Comment by leilani-reich** [comment]
Hi Carlos, after seeing Malia's parser function and reviewing parser.R and download_jira_data.Rmd, I was wondering, should the downloader just return the paths to the saved json files (so there would be two paths in this case, one for issues, and one for comments) instead of a json object?

**Comment by leilani-reich** [comment]
Hi @carlosparadis,

the first issue comment has been updated with tasks we assigned for functions & commands to wrap around. We tried to assign ourselves tasks that were not dependent on each other. Please let us know if anything should be changed. Thank you.

**Comment by carlosparadis** [comment]
@leilani-reich is this the right issue? Or you meant DV8?

**Comment by leilani-reich** [comment]
Sorry, I meant to put this under DV8.

**Comment by carlosparadis** [comment]
@leilani-reich 

Please go ahead and make a pull request of the code I will paste here, as it includes a few changes. But there are some critical things missing. It seems you have not included pagination. Which is to say, if I query 2023 and beyond, your function download the first 20 issues...and misses all the rest. 

The code is this:

```r
#'  Download Bugzilla issues and comments using the Bugzilla REST API
#'
#' @param bugzilla_site link to specific bugzilla site
#' @param start_timestamp when to start bug retrieval (ex. 2023-01-01T00:14:57Z)
#' @param end_timestamp when to end bug retrieval (ex. 2023-01-01T00:14:57Z)
#' @param project_name the name of the bugzilla project (ex. redhat for https://bugzilla.redhat.com")
#' @param comments whether to download comments
#' @export
download_bugzilla_from_rest_api <- function(bugzilla_site, start_timestamp, end_timestamp, project_name, comments=FALSE){
  # Format link to retrieve data using Bugzilla REST API
  bugzilla_site <- paste(bugzilla_site, "/rest", sep="")

  # Get the issues starting from the specified date
  issues <- httr::GET(paste(bugzilla_site, "/bug", "?creation_time=", start_timestamp, sep=""))
  json_issues <- rawToChar(issues$content)

  # Get table for the bug issues
  bug_issue_table <- jsonlite::parse_json(json_issues)
  # There will likely be a loop here for pagination.
  # You will want to save each page to disk as they come. That way
  # your code can recover without re-downloading if something goes wrong.
  # For now, just include pagination and save the file at every step of the
  # loop with different names (the page number would likely be the differentiator
  # in name).
  # In posession of Malia's parser function, someone can then parse the folder
  # to know what timestamp to resume from in case of error.


  # Get the bug ids. Needed to know which comments to retrieve.
  bug_ids <- sapply(bug_issue_table[["bugs"]],function(x) x[["id"]])

  return(bug_ids)


  ## This should be moved to a separate function, that take as input bug-ids,
  # and the other necessary parameters to download the code.

  json_comments <- list()
  if (comments == TRUE){
    # Get the comments associated with the bugs acquired
    for (i in 1:length(bug_ids)){
      comments <- httr::GET(paste(bugzilla_site, "/bug/", bug_ids[i], "/comment", sep=""))


      json_comments[[i]] <- jsonlite::parse_json(rawToChar(comments$content))

      # The code should be expanded here to save to disk json_comments[[i]] before
      # the next step. Imagine 10k comments are downloaded over night.
      # And comment 9k fails. Rather than re-run 9k GET requests again,
      # we now only need to re-run 1k. It substantially increases the number of
      # I/Os but it makes it much more failsafe.

    }
  }


}
```


Some of the changes I made and why are:

## 1. Stick to as close to raw data as possible

> #bug_issue_table <- data.table(jsonlite::stream_in(textConnection(json_issues),verbose = FALSE)) 

Was replaced by:

bug_issue_table <- jsonlite::parse_json(json_issues)

On a downloader, you want to pre-process as minimally as possible. This will continue to represent the data as a list, which is R's closest representation to json. It is ok to extract and tabulate information that you may need for the processing, however the saved file must still be the bug_issue_table above with minimal to none pre-processing. 

In the list above, you can extract the ids to an array with the following 1 line:

`bug_ids <- sapply(bug_issue_table[["bugs"]],function(x) x[["id"]])`

## 2. Pagination is missing

I believe your function is missing pagination. When you make a GET request, or when you access issues on a browser, it normally fetches a few of them per page, and you must iterate through the pages in order to obtain all the issues. Here's the equivalent search as you included examples for me:

https://bugzilla.redhat.com/buglist.cgi?chfieldfrom=2023-01-01&chfieldto=Now&columnlist=product%2Ccomponent%2Cassigned_to%2Cstatus%2Csummary%2Clast_change_time%2Cseverity%2Cpriority&order=priority%2C%20severity%2C%20&query_format=advanced 

As you can see, at the bottom it states:

> Showing 1 to 20 of 38,501 entriesNo rows selected

Moreover, with your code example I see:

<img width="539" alt="Screen Shot 2023-03-23 at 10 56 59 PM" src="https://user-images.githubusercontent.com/17270563/227436871-3d9baf3b-9c2a-4361-a795-c99c06a18578.png">

## 3. De-coupling

I feel hiding the filtering within the function can be misleading behavior to the user (my first assumption would be that I am only downloading issues for the specified time range, when in fact this is not happening and I may be hitting the server with a much larger request). 

To avoid confusion, de-couple the issue and comment functions. 

## 4. Failsafe

Since we can't control the enddate parameter, it is best we save as we go, in case issue 9999 out of 10k fails, we don't need to re-download everything again.

Also, bug https://bugzilla.redhat.com/show_bug.cgi?id=2181187 is a helpful example for Malia as it contain comments. So I will forward to her in her issue.

**Comment by carlosparadis** [comment]
I've explained a bit more my comments on #166 to Malia. Please have a look at that issue, and let me know if you have questions.

**Comment by carlosparadis** [comment]
This issue is remaining open until the Unit tests are done, but otherwise all other tasks should open a new issue in case of a problem as their associated pull requests are already merged or about to be merged to Kaiaulu's master branch.
